File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/StorageOptions.java
package au.com.shiftyjelly.pocketcasts.repositories.file;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.Resources;
import android.os.Environment;

import java.io.File;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import au.com.shiftyjelly.pocketcasts.localization.R;

public class StorageOptions {
	
	private ArrayList<FolderLocation> folderLocations = null;

	public List<FolderLocation> getFolderLocations(Context context){
		if (folderLocations == null){
			folderLocations = new ArrayList<>();

            ArrayList<String> confirmedMountPoints = null;
			File[] externalDirs = context.getExternalFilesDirs(null);

			confirmedMountPoints = new ArrayList<>();
			if (externalDirs != null && externalDirs.length > 0){
				for (File directory : externalDirs){
					if (directory != null) {
						confirmedMountPoints.add(directory.getAbsolutePath());
					}
				}
			}

            testAndCleanMountsList(confirmedMountPoints);
            determineFolderLocations(confirmedMountPoints, context);
		}
		
		return folderLocations;
	}

	private void testAndCleanMountsList(ArrayList<String> confirmedMountPoints) {
		/*
		 * Now that we have a cleaned list of mount paths Test each one to make
		 * sure it's a valid and available path. If it is not, remove it from
		 * the list.
		 */

		for (Iterator<String> it = confirmedMountPoints.iterator(); it.hasNext();){
			String mount = it.next();
			File folder = new File(mount);
			if (!folder.exists() || !folder.isDirectory() || !folder.canWrite()){
				it.remove();
			}
		}
	}

	@SuppressLint("NewApi")
	private void determineFolderLocations(ArrayList<String> confirmedMountPoints, Context context) {
		if (confirmedMountPoints.size() == 0) return;
		
		int externalSDCardCount = 1;
		String firstMountPoint = confirmedMountPoints.get(0);
		confirmedMountPoints.remove(0);

		Resources resources = context.getResources();
		//the first mount point is different to the rest
		if (!Environment.isExternalStorageRemovable() || Environment.isExternalStorageEmulated()){
			folderLocations.add(new FolderLocation(firstMountPoint, resources.getString(R.string.settings_storage_phone)));
		}
		else {
			folderLocations.add(new FolderLocation(firstMountPoint, resources.getString(R.string.settings_storage_sd_card)));
			externalSDCardCount++;
		}
		
		//label all the rest as external storage
		for (String mountPoint : confirmedMountPoints){
			String label = externalSDCardCount == 1 ? resources.getString(R.string.settings_storage_sd_card) : resources.getString(R.string.settings_storage_sd_card_number, String.valueOf(externalSDCardCount));
			folderLocations.add(new FolderLocation(mountPoint, label));
			externalSDCardCount++;
		}
	}
}

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/StorageOptions.kt
package au.com.shiftyjelly.pocketcasts.repositories.file

import android.content.Context
import android.os.Environment
import au.com.shiftyjelly.pocketcasts.localization.R
import java.io.File

class StorageOptions {

    private var folderLocations: ArrayList<FolderLocation>? = null

    fun getFolderLocations(context: Context): List<FolderLocation> {
        if (folderLocations == null) {
            folderLocations = arrayListOf()
            val confirmedMountPoints = arrayListOf<String>()

            val externalDirs = context.getExternalFilesDirs(null)
            if (externalDirs != null && externalDirs.isNotEmpty()) {
                for (directory in externalDirs) {
                    if (directory != null) {
                        confirmedMountPoints.add(directory.absolutePath)
                    }
                }
            }

            testAndCleanMountsList(confirmedMountPoints)
            determineFolderLocations(confirmedMountPoints, context)
        }

        return folderLocations as List<FolderLocation>
    }

    private fun testAndCleanMountsList(confirmedMountPoints: ArrayList<String>) {
        /*
		 * Now that we have a cleaned list of mount paths Test each one to make
		 * sure it's a valid and available path. If it is not, remove it from
		 * the list.
		 */
        val iterator = confirmedMountPoints.iterator()
        while (iterator.hasNext()) {
            val mount = iterator.next()
            val folder = File(mount)
            if (!folder.exists() || !folder.isDirectory || !folder.canWrite()) {
                iterator.remove()
            }
        }
    }

    private fun determineFolderLocations(
        confirmedMountPoints: ArrayList<String>,
        context: Context
    ) {
        if (confirmedMountPoints.isEmpty()) return

        var externalSDCardCount = 1
        val firstMountPoint = confirmedMountPoints[0]
        confirmedMountPoints.removeAt(0)

        val resources = context.resources
        // the first mount point is different to the rest
        if (!Environment.isExternalStorageRemovable() ||
            Environment.isExternalStorageEmulated()
        ) {
            requireNotNull(folderLocations) {
                "folderLocations can not be null"
            }.add(
                FolderLocation(
                    firstMountPoint,
                    resources.getString(R.string.settings_storage_phone)
                )
            )
        } else {
            requireNotNull(folderLocations) {
                "folderLocations can not be null"
            }.add(
                FolderLocation(
                    firstMountPoint,
                    resources.getString(R.string.settings_storage_sd_card)
                )
            )
            externalSDCardCount++
        }

        // label all the rest as external storage
        for (mountPoint in confirmedMountPoints) {
            val label = if (externalSDCardCount == 1) {
                resources.getString(R.string.settings_storage_sd_card)
            } else {
                resources.getString(
                    R.string.settings_storage_sd_card_number,
                    externalSDCardCount.toString()
                )
            }
            requireNotNull(folderLocations) {
                "folderLocations can not be null"
            }.add(FolderLocation(mountPoint, label))
            externalSDCardCount++
        }
    }
}
